### PR TITLE
[PM-22018]: Harden arithmetic using checked operations in single_tx.rs

### PR DESCRIPTION
**Pull request by m2ux**
## Summary

Replace unchecked arithmetic operations in `single_tx.rs` with checked alternatives (`checked_mul`, `checked_sub`, `checked_add`) and proper error propagation, converting panics on overflow/underflow into recoverable errors. Addresses informational finding from the Least Authority Node DIFF Audit (Feb 2026).

🎫 [Ticket](https://shielded.atlassian.net/browse/PM-22018)  📐 [Engineering](https://github.com/shieldedtech/midnight-agent-eng/blob/mike/.engineering/artifacts/planning/2026-04-13-harden-arithmetic-checked-operations/README.md)

---

## Motivation

The transaction builder in `util/toolkit/src/tx_generator/builder/builders/single_tx.rs` uses `checked_mul` for amount aggregation but calls `.expect()` on the result, converting arithmetic overflow into a panic rather than a recoverable error. Similarly, subtraction for change calculation may underflow with adversarial inputs. While the toolkit is not consensus-critical, panics in production tooling are undesirable and violate defence-in-depth principles.

This change ensures that large or adversarial input values produce typed errors that callers can handle gracefully, rather than crashing the process.

---

## Changes

**Implementation:**
- Replace `.expect()` calls on `checked_mul` results with proper error propagation via `?` / `ok_or()`
- Guard subtraction underflow paths with checked operations and error returns
- Extend existing error types (`ShieldedCoinSelectionError`, `UtxoSelectionError`) with arithmetic overflow/underflow variants
- Add unit tests covering boundary values (u128::MAX, 0, values near overflow thresholds)

---

## 📌 Submission Checklist

- [ ] Changes are backward-compatible (or flagged if breaking)
- [ ] Pull request description explains why the change is needed
- [x] Self-reviewed the diff
- [ ] I have included a change file, or skipped for this reason: [reason]
- [ ] If the changes introduce a new feature, I have bumped the node minor version
- [ ] Update documentation (if relevant)
- [ ] No new todos introduced

---

## 🔱 Fork Strategy

- [ ] Node Runtime Update
- [ ] Node Client Update
- [ ] Other
- [x] N/A

---

## 🗹 TODO before merging

- [ ] Ready for review

### CODE REVIEW

File: ledger/helpers/src/versions/common/input.rs [hunk 1]
```diff
@@ -20,6 +20,8 @@ use std::sync::Arc;
 
 #[derive(Debug, thiserror::Error)]
 pub enum ShieldedCoinSelectionError {
+	#[error("arithmetic overflow in shielded coin selection")]
+	ArithmeticOverflow,
```
Comment (gilescope) [inline ledger/helpers/src/versions/common/input.rs]: for backward compatibility shall we add this at the end of the enum?

**Review by gilescope — APPROVED** [review]

File: ledger/helpers/src/versions/common/input.rs (hunk 1 above)
Comment (m2ux) [reply ledger/helpers/src/versions/common/input.rs]: Good call — moved `ArithmeticOverflow` to the end of `ShieldedCoinSelectionError` (and applied the same to `UtxoSelectionError`) in 9c17350. Existing variant discriminants are now preserved.
